[Bug]: Batch proposals in graphQL not working

### Problem encountered

In the overnight run the test `test_graphQl_CheckMetaData` is failing because a grapqhQL call that list proposals in struggling when a batch proposal exists:
```
            for prop in json_data["data"]["proposalsConnection"]["edges"]:
>               prop_node = prop["node"]
E               TypeError: 'NoneType' object is not subscriptable
```

Running a query through the graphQL playground on the same data gives me:
<img width="1627" alt="image" src="https://github.com/vegaprotocol/vega/assets/1857660/d1ca1261-e961-4941-88d8-d5c2fb1d7a04">

which is likely because in the graphQL schema the `terms` field is non-nullable but I guess is empty when its a batch proposal.

Also a field for `batchTerms` seems to be missing from the graphQL type, but exists in the gRPC endpoint.

### Observed behaviour

GraphQL is not happy querying batch proposals.

### Expected behaviour

It should be possible to query batch proposals, and the batched terms, through graphQL

### Steps to reproduce

```bash
1. Submit `xyz` transaction
2. With this payload
3. Query `abc` API
4. Observe the error
```


### Software version

develop

### Failing test

_No response_

### Jenkins run

_No response_

### Configuration used

_No response_

### Relevant log output

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -116,6 +116,7 @@
 - [10354](https://github.com/vegaprotocol/vega/issues/10354) - Renumbered SQL migration scripts 0055-0067 due to 0068 being released as part of a patch without renumbering.
 - [476](https://github.com/vegaprotocol/core-test-coverage/issues/476) - Add tests for markets expiring in opening auction, fix a bug for future markets.
 - [10339](https://github.com/vegaprotocol/vega/issues/10339) - Fix graphQL support for batch proposals.
+- [10339](https://github.com/vegaprotocol/vega/issues/10339) - Fix for GraphQL batch proposals support.
 
 ## 0.73.0
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -112,8 +112,10 @@
 - [10321](https://github.com/vegaprotocol/vega/issues/10321) - Fix `ListPartyMarginModes` panic.
 - [10324](https://github.com/vegaprotocol/vega/issues/10324) - Fix `GetMarketHistoryByID` to only return the most recent market data information when no dates are provided.
 - [10318](https://github.com/vegaprotocol/vega/issues/10318) - Prevent stop orders being placed during opening auction.
-- [476](https://github.com/vegaprotocol/core-test-coverage/issues/476) -  Add tests for markets expiring in opening auction, fix a bug for future markets.
+- [476](https://github.com/vegaprotocol/core-test-coverage/issues/476) - Add tests for markets expiring in opening auction, fix a bug for future markets.
 - [10354](https://github.com/vegaprotocol/vega/issues/10354) - Renumbered SQL migration scripts 0055-0067 due to 0068 being released as part of a patch without renumbering.
+- [476](https://github.com/vegaprotocol/core-test-coverage/issues/476) - Add tests for markets expiring in opening auction, fix a bug for future markets.
+- [10339](https://github.com/vegaprotocol/vega/issues/10339) - Fix graphQL support for batch proposals.
 
 ## 0.73.0
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -115,7 +115,6 @@
 - [476](https://github.com/vegaprotocol/core-test-coverage/issues/476) - Add tests for markets expiring in opening auction, fix a bug for future markets.
 - [10354](https://github.com/vegaprotocol/vega/issues/10354) - Renumbered SQL migration scripts 0055-0067 due to 0068 being released as part of a patch without renumbering.
 - [476](https://github.com/vegaprotocol/core-test-coverage/issues/476) - Add tests for markets expiring in opening auction, fix a bug for future markets.
-- [10339](https://github.com/vegaprotocol/vega/issues/10339) - Fix graphQL support for batch proposals.
 - [10339](https://github.com/vegaprotocol/vega/issues/10339) - Fix for GraphQL batch proposals support.
 
 ## 0.73.0

--- a/datanode/api/trading_data_v2.go
+++ b/datanode/api/trading_data_v2.go
@@ -2309,6 +2309,10 @@ func (t *TradingDataServiceV2) ListGovernanceData(ctx context.Context, req *v2.L
 		if err != nil {
 			return nil, formatE(ErrGovernanceServiceGetVotes, errors.Wrapf(err, "proposalID: %s", proposalID))
 		}
+
+		if len(edges[i].Node.Proposals) > 0 {
+			edges[i].Node.ProposalType = vega.GovernanceData_TYPE_BATCH
+		}
 	}
 
 	proposalsConnection := &v2.GovernanceDataConnection{

--- a/datanode/gateway/graphql/gqlgen.yml
+++ b/datanode/gateway/graphql/gqlgen.yml
@@ -279,6 +279,10 @@ models:
     model: code.vegaprotocol.io/vega/protos/vega.ProposalRationale
   ProposalTerms:
     model: code.vegaprotocol.io/vega/protos/vega.ProposalTerms
+  BatchProposalTerms:
+    model: code.vegaprotocol.io/vega/protos/vega.BatchProposalTerms
+  BatchProposalTermsChange:
+    model: code.vegaprotocol.io/vega/protos/vega.BatchProposalTermsChange
   ProposalDetail:
     model: code.vegaprotocol.io/vega/protos/vega.Proposal
   NewMarket:

--- a/datanode/gateway/graphql/market_resolvers.go
+++ b/datanode/gateway/graphql/market_resolvers.go
@@ -248,7 +248,7 @@ func (r *myMarketResolver) LiquidityMonitoringParameters(_ context.Context, obj 
 	}, nil
 }
 
-func (r *myMarketResolver) Proposal(ctx context.Context, obj *types.Market) (ProposalNode, error) {
+func (r *myMarketResolver) MarketProposal(ctx context.Context, obj *types.Market) (ProposalNode, error) {
 	resp, err := r.tradingDataClientV2.GetGovernanceData(ctx, &v2.GetGovernanceDataRequest{
 		ProposalId: &obj.Id,
 	})
@@ -262,6 +262,20 @@ func (r *myMarketResolver) Proposal(ctx context.Context, obj *types.Market) (Pro
 	resolver := (*proposalEdgeResolver)(r)
 	if resp.GetData().ProposalType == vega.GovernanceData_TYPE_BATCH {
 		return resolver.BatchProposal(ctx, resp.GetData())
+	}
+
+	return resp.Data, nil
+}
+
+func (r *myMarketResolver) Proposal(ctx context.Context, obj *types.Market) (*vega.GovernanceData, error) {
+	resp, err := r.tradingDataClientV2.GetGovernanceData(ctx, &v2.GetGovernanceDataRequest{
+		ProposalId: &obj.Id,
+	})
+	// it's possible to not find a proposal as of now.
+	// some market are loaded at startup, without
+	// going through the proposal phase
+	if err != nil {
+		return nil, nil //nolint:nilerr
 	}
 
 	return resp.Data, nil

--- a/datanode/gateway/graphql/market_resolvers.go
+++ b/datanode/gateway/graphql/market_resolvers.go
@@ -248,7 +248,7 @@ func (r *myMarketResolver) LiquidityMonitoringParameters(_ context.Context, obj 
 	}, nil
 }
 
-func (r *myMarketResolver) Proposal(ctx context.Context, obj *types.Market) (*types.GovernanceData, error) {
+func (r *myMarketResolver) Proposal(ctx context.Context, obj *types.Market) (ProposalNode, error) {
 	resp, err := r.tradingDataClientV2.GetGovernanceData(ctx, &v2.GetGovernanceDataRequest{
 		ProposalId: &obj.Id,
 	})
@@ -258,6 +258,12 @@ func (r *myMarketResolver) Proposal(ctx context.Context, obj *types.Market) (*ty
 	if err != nil {
 		return nil, nil //nolint:nilerr
 	}
+
+	resolver := (*proposalEdgeResolver)(r)
+	if resp.GetData().ProposalType == vega.GovernanceData_TYPE_BATCH {
+		return resolver.BatchProposal(ctx, resp.GetData())
+	}
+
 	return resp.Data, nil
 }
 

--- a/datanode/gateway/graphql/models.go
+++ b/datanode/gateway/graphql/models.go
@@ -118,7 +118,7 @@ type AuctionDuration struct {
 }
 
 type BatchProposal struct {
-	// Proposal ID that is filled by Vega once proposal reaches the network
+	// Proposal ID that is provided by Vega once proposal reaches the network
 	ID *string `json:"id"`
 	// A UUID reference to aid tracking proposals on Vega
 	Reference string `json:"reference"`
@@ -126,25 +126,25 @@ type BatchProposal struct {
 	Party *vega.Party `json:"party"`
 	// State of the proposal
 	State vega.Proposal_State `json:"state"`
-	// RFC3339Nano time and date when the proposal reached Vega network
+	// RFC3339Nano time and date when the proposal reached the network
 	Datetime int64 `json:"datetime"`
 	// Rationale behind the proposal
 	Rationale *vega.ProposalRationale `json:"rationale"`
 	// Votes cast for this proposal
 	Votes *ProposalVotes `json:"votes"`
-	// Why the proposal was rejected by the core
+	// Reason the proposal was rejected
 	RejectionReason *vega.ProposalError `json:"rejectionReason"`
-	// Error details of the rejectionReason
+	// Details of the rejection reason
 	ErrorDetails *string `json:"errorDetails"`
 	// Required majority for this proposal to succeed
 	RequiredMajority string `json:"requiredMajority"`
 	// Required participation for this proposal to succeed
 	RequiredParticipation string `json:"requiredParticipation"`
-	// Equity-like share required for a market amendment proposal to be enacted (if not met, the proposal will not be enacted), represented as a fraction that can be converted to a percentage
+	// Equity-like share required for a market amendment proposal to be enacted, represented as a fraction that can be converted to a percentage. If not met, the proposal will not be enacted
 	RequiredLpMajority *string `json:"requiredLpMajority"`
-	// The market share of LPs' equity-like share that must take part in the market amendment vote for the proposal to pass. This means the votes of LPs that have submitted more liquidity to that market, or have been LPs from the start carry more weight. If it requires 50% of a market's equity-like share for a majority, and the proposal receives only YES votes but only LPs with 49% of the equity-like share voted, the proposal will not pass
+	// The market share of LPs' equity-like share that must take part in a market amendment vote for the proposal to pass. This means the votes of LPs that have submitted more liquidity to that market, or have been LPs from the start carry more weight. If it requires 50% of a market's equity-like share for a majority, and the full batch of proposals receives all YES votes but only LPs with 49% of the equity-like share voted, the proposal will not pass
 	RequiredLpParticipation *string `json:"requiredLpParticipation"`
-	// Child proposals that are part of the batch
+	// Proposals that are part of the batch
 	SubProposals []*vega.Proposal `json:"subProposals"`
 }
 

--- a/datanode/gateway/graphql/models.go
+++ b/datanode/gateway/graphql/models.go
@@ -62,6 +62,10 @@ type ProposalChange interface {
 	IsProposalChange()
 }
 
+type ProposalNode interface {
+	IsProposalNode()
+}
+
 type RiskModel interface {
 	IsRiskModel()
 }
@@ -112,6 +116,39 @@ type AuctionDuration struct {
 	// Target uncrossing trading volume
 	Volume int `json:"volume"`
 }
+
+type BatchProposal struct {
+	// Proposal ID that is filled by Vega once proposal reaches the network
+	ID *string `json:"id"`
+	// A UUID reference to aid tracking proposals on Vega
+	Reference string `json:"reference"`
+	// Party that prepared the proposal
+	Party *vega.Party `json:"party"`
+	// State of the proposal
+	State vega.Proposal_State `json:"state"`
+	// RFC3339Nano time and date when the proposal reached Vega network
+	Datetime int64 `json:"datetime"`
+	// Rationale behind the proposal
+	Rationale *vega.ProposalRationale `json:"rationale"`
+	// Votes cast for this proposal
+	Votes *ProposalVotes `json:"votes"`
+	// Why the proposal was rejected by the core
+	RejectionReason *vega.ProposalError `json:"rejectionReason"`
+	// Error details of the rejectionReason
+	ErrorDetails *string `json:"errorDetails"`
+	// Required majority for this proposal to succeed
+	RequiredMajority string `json:"requiredMajority"`
+	// Required participation for this proposal to succeed
+	RequiredParticipation string `json:"requiredParticipation"`
+	// Equity-like share required for a market amendment proposal to be enacted (if not met, the proposal will not be enacted), represented as a fraction that can be converted to a percentage
+	RequiredLpMajority *string `json:"requiredLpMajority"`
+	// The market share of LPs' equity-like share that must take part in the market amendment vote for the proposal to pass. This means the votes of LPs that have submitted more liquidity to that market, or have been LPs from the start carry more weight. If it requires 50% of a market's equity-like share for a majority, and the proposal receives only YES votes but only LPs with 49% of the equity-like share voted, the proposal will not pass
+	RequiredLpParticipation *string `json:"requiredLpParticipation"`
+	// Child proposals that are part of the batch
+	SubProposals []*vega.Proposal `json:"subProposals"`
+}
+
+func (BatchProposal) IsProposalNode() {}
 
 // A Vega builtin asset, mostly for testing purpose
 type BuiltinAsset struct {

--- a/datanode/gateway/graphql/proposa_edge_resolver.go
+++ b/datanode/gateway/graphql/proposa_edge_resolver.go
@@ -37,6 +37,14 @@ func (r *proposalEdgeResolver) ProposalNode(ctx context.Context, data *v2.Govern
 	return data.Node, nil
 }
 
+func (r *proposalEdgeResolver) Proposal(ctx context.Context, data *v2.GovernanceDataEdge) (*vega.GovernanceData, error) {
+	if data == nil || data.Node == nil || data.Node.Proposal == nil {
+		return nil, ErrInvalidProposal
+	}
+
+	return data.Node, nil
+}
+
 func (r *proposalEdgeResolver) BatchProposal(ctx context.Context, data *types.GovernanceData) (ProposalNode, error) {
 	proposal := data.Proposal
 

--- a/datanode/gateway/graphql/proposa_edge_resolver.go
+++ b/datanode/gateway/graphql/proposa_edge_resolver.go
@@ -1,0 +1,71 @@
+// Copyright (C) 2023 Gobalsky Labs Limited
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package gql
+
+import (
+	"context"
+
+	v2 "code.vegaprotocol.io/vega/protos/data-node/api/v2"
+	types "code.vegaprotocol.io/vega/protos/vega"
+	vega "code.vegaprotocol.io/vega/protos/vega"
+)
+
+type proposalEdgeResolver VegaResolverRoot
+
+func (r *proposalEdgeResolver) Node(ctx context.Context, data *v2.GovernanceDataEdge) (ProposalNode, error) {
+	if data == nil || data.Node == nil || data.Node.Proposal == nil {
+		return nil, ErrInvalidProposal
+	}
+
+	if data.GetNode().ProposalType == vega.GovernanceData_TYPE_BATCH {
+		return r.BatchProposal(ctx, data.Node)
+	}
+
+	return data.Node, nil
+}
+
+func (r *proposalEdgeResolver) BatchProposal(ctx context.Context, data *types.GovernanceData) (ProposalNode, error) {
+	proposal := data.Proposal
+
+	resolver := (*proposalResolver)(r)
+
+	party, err := resolver.Party(ctx, data)
+	if err != nil {
+		return nil, err
+	}
+
+	votes, err := resolver.Votes(ctx, data)
+	if err != nil {
+		return nil, err
+	}
+
+	return BatchProposal{
+		ID:                      &proposal.Id,
+		Reference:               proposal.Reference,
+		Party:                   party,
+		State:                   proposal.State,
+		Datetime:                proposal.Timestamp,
+		Rationale:               proposal.Rationale,
+		Votes:                   votes,
+		RejectionReason:         proposal.Reason,
+		ErrorDetails:            proposal.ErrorDetails,
+		RequiredMajority:        proposal.RequiredMajority,
+		RequiredParticipation:   proposal.RequiredParticipation,
+		RequiredLpMajority:      proposal.RequiredLiquidityProviderMajority,
+		RequiredLpParticipation: proposal.RequiredLiquidityProviderParticipation,
+		SubProposals:            data.Proposals,
+	}, nil
+}

--- a/datanode/gateway/graphql/proposa_edge_resolver.go
+++ b/datanode/gateway/graphql/proposa_edge_resolver.go
@@ -25,7 +25,7 @@ import (
 
 type proposalEdgeResolver VegaResolverRoot
 
-func (r *proposalEdgeResolver) Node(ctx context.Context, data *v2.GovernanceDataEdge) (ProposalNode, error) {
+func (r *proposalEdgeResolver) ProposalNode(ctx context.Context, data *v2.GovernanceDataEdge) (ProposalNode, error) {
 	if data == nil || data.Node == nil || data.Node.Proposal == nil {
 		return nil, ErrInvalidProposal
 	}

--- a/datanode/gateway/graphql/proposal_edge_resolver.go
+++ b/datanode/gateway/graphql/proposal_edge_resolver.go
@@ -42,6 +42,10 @@ func (r *proposalEdgeResolver) Proposal(ctx context.Context, data *v2.Governance
 		return nil, ErrInvalidProposal
 	}
 
+	if data.GetNode().ProposalType == vega.GovernanceData_TYPE_BATCH {
+		return nil, nil
+	}
+
 	return data.Node, nil
 }
 

--- a/datanode/gateway/graphql/proposal_terms_resolver.go
+++ b/datanode/gateway/graphql/proposal_terms_resolver.go
@@ -22,9 +22,61 @@ import (
 
 	"code.vegaprotocol.io/vega/libs/ptr"
 	types "code.vegaprotocol.io/vega/protos/vega"
+	vega "code.vegaprotocol.io/vega/protos/vega"
 )
 
 var ErrUnsupportedProposalTermsChanges = errors.New("unsupported proposal terms changes")
+
+type batchProposalTermsResolver VegaResolverRoot
+
+func (r *batchProposalTermsResolver) ClosingDatetime(ctx context.Context, obj *vega.BatchProposalTerms) (int64, error) {
+	// this is a unix timestamp (specified by users)
+	// needs to convert to time then UnixNano for the Timestamp resolver to work
+	return time.Unix(obj.ClosingTimestamp, 0).UnixNano(), nil
+}
+
+type batchProposalTermsChangeResolver VegaResolverRoot
+
+func (r *batchProposalTermsChangeResolver) EnactmentDatetime(ctx context.Context, obj *vega.BatchProposalTermsChange) (*int64, error) {
+	var dt *int64
+	if obj.EnactmentTimestamp != 0 {
+		// this is a unix timestamp (specified by users)
+		// needs to convert to time then UnixNano for the Timestamp resolver to work
+		dt = ptr.From(time.Unix(obj.EnactmentTimestamp, 0).UnixNano())
+	}
+	return dt, nil
+}
+
+func (r *batchProposalTermsChangeResolver) Change(ctx context.Context, obj *vega.BatchProposalTermsChange) (ProposalChange, error) {
+	switch obj.Change.(type) {
+	case *types.BatchProposalTermsChange_UpdateMarket:
+		return obj.GetUpdateMarket(), nil
+	case *types.BatchProposalTermsChange_UpdateNetworkParameter:
+		return obj.GetUpdateNetworkParameter(), nil
+	case *types.BatchProposalTermsChange_NewMarket:
+		return obj.GetNewMarket(), nil
+	case *types.BatchProposalTermsChange_NewFreeform:
+		return obj.GetNewFreeform(), nil
+	case *types.BatchProposalTermsChange_UpdateAsset:
+		return obj.GetUpdateAsset(), nil
+	case *types.BatchProposalTermsChange_CancelTransfer:
+		return obj.GetCancelTransfer(), nil
+	case *types.BatchProposalTermsChange_NewTransfer:
+		return obj.GetNewTransfer(), nil
+	case *types.BatchProposalTermsChange_NewSpotMarket:
+		return obj.GetNewSpotMarket(), nil
+	case *types.BatchProposalTermsChange_UpdateSpotMarket:
+		return obj.GetUpdateSpotMarket(), nil
+	case *types.BatchProposalTermsChange_UpdateMarketState:
+		return obj.GetUpdateMarketState(), nil
+	case *types.BatchProposalTermsChange_UpdateReferralProgram:
+		return obj.GetUpdateReferralProgram(), nil
+	case *types.BatchProposalTermsChange_UpdateVolumeDiscountProgram:
+		return obj.GetUpdateVolumeDiscountProgram(), nil
+	default:
+		return nil, ErrUnsupportedProposalTermsChanges
+	}
+}
 
 type proposalTermsResolver VegaResolverRoot
 

--- a/datanode/gateway/graphql/resolvers.go
+++ b/datanode/gateway/graphql/resolvers.go
@@ -191,6 +191,11 @@ func (r *VegaResolverRoot) Proposal() ProposalResolver {
 	return (*proposalResolver)(r)
 }
 
+// Proposal returns the proposal resolver.
+func (r *VegaResolverRoot) ProposalEdge() ProposalEdgeResolver {
+	return (*proposalEdgeResolver)(r)
+}
+
 // ProposalDetail returns the Proposal detail resolver.
 func (r *VegaResolverRoot) ProposalDetail() ProposalDetailResolver {
 	return (*proposalDetailResolver)(r)
@@ -266,6 +271,14 @@ func (r *VegaResolverRoot) NewMarket() NewMarketResolver {
 
 func (r *VegaResolverRoot) ProposalTerms() ProposalTermsResolver {
 	return (*proposalTermsResolver)(r)
+}
+
+func (r *VegaResolverRoot) BatchProposalTerms() BatchProposalTermsResolver {
+	return (*batchProposalTermsResolver)(r)
+}
+
+func (r *VegaResolverRoot) BatchProposalTermsChange() BatchProposalTermsChangeResolver {
+	return (*batchProposalTermsChangeResolver)(r)
 }
 
 func (r *VegaResolverRoot) UpdateMarket() UpdateMarketResolver {

--- a/datanode/gateway/graphql/resolvers.go
+++ b/datanode/gateway/graphql/resolvers.go
@@ -1428,7 +1428,6 @@ func (r *myQueryResolver) ProposalsConnection(ctx context.Context, proposalType 
 
 func (r *myQueryResolver) Proposal(ctx context.Context, id *string, reference *string) (ProposalNode, error) {
 	if id != nil {
-
 		resp, err := r.tradingDataClientV2.GetGovernanceData(ctx, &v2.GetGovernanceDataRequest{
 			ProposalId: id,
 		})

--- a/datanode/gateway/graphql/resolvers.go
+++ b/datanode/gateway/graphql/resolvers.go
@@ -1426,14 +1426,21 @@ func (r *myQueryResolver) ProposalsConnection(ctx context.Context, proposalType 
 	return handleProposalsRequest(ctx, r.tradingDataClientV2, nil, nil, proposalType, inState, pagination)
 }
 
-func (r *myQueryResolver) Proposal(ctx context.Context, id *string, reference *string) (*vegapb.GovernanceData, error) {
+func (r *myQueryResolver) Proposal(ctx context.Context, id *string, reference *string) (ProposalNode, error) {
 	if id != nil {
+
 		resp, err := r.tradingDataClientV2.GetGovernanceData(ctx, &v2.GetGovernanceDataRequest{
 			ProposalId: id,
 		})
 		if err != nil {
 			return nil, err
 		}
+
+		resolver := (*proposalEdgeResolver)(r)
+		if resp.GetData().ProposalType == vega.GovernanceData_TYPE_BATCH {
+			return resolver.BatchProposal(ctx, resp.GetData())
+		}
+
 		return resp.Data, nil
 	} else if reference != nil {
 		resp, err := r.tradingDataClientV2.GetGovernanceData(ctx, &v2.GetGovernanceDataRequest{
@@ -1442,6 +1449,12 @@ func (r *myQueryResolver) Proposal(ctx context.Context, id *string, reference *s
 		if err != nil {
 			return nil, err
 		}
+
+		resolver := (*proposalEdgeResolver)(r)
+		if resp.GetData().ProposalType == vega.GovernanceData_TYPE_BATCH {
+			return resolver.BatchProposal(ctx, resp.GetData())
+		}
+
 		return resp.Data, nil
 	}
 

--- a/datanode/gateway/graphql/resolvers_test.go
+++ b/datanode/gateway/graphql/resolvers_test.go
@@ -671,7 +671,6 @@ func TestNewResolverRoot_Proposals(t *testing.T) {
 		)
 
 		var (
-			p         *protoTypes.GovernanceData
 			terms     *protoTypes.ProposalTerms
 			newMarket *protoTypes.ProposalTerms_NewMarket
 			asset     *protoTypes.Asset
@@ -679,7 +678,8 @@ func TestNewResolverRoot_Proposals(t *testing.T) {
 			err       error
 		)
 
-		p, err = root.Query().Proposal(ctx, &id, nil)
+		pn, err := root.Query().Proposal(ctx, &id, nil)
+		p := pn.(*protoTypes.GovernanceData)
 
 		t.Run("Proposal terms should be for a new market", func(t *testing.T) {
 			terms, err = root.Proposal().Terms(ctx, p)
@@ -731,7 +731,6 @@ func TestNewResolverRoot_Proposals(t *testing.T) {
 		)
 
 		var (
-			p         *protoTypes.GovernanceData
 			terms     *protoTypes.ProposalTerms
 			newMarket *protoTypes.ProposalTerms_NewSpotMarket
 			asset     *protoTypes.Asset
@@ -739,7 +738,8 @@ func TestNewResolverRoot_Proposals(t *testing.T) {
 			err       error
 		)
 
-		p, err = root.Query().Proposal(ctx, &id, nil)
+		pn, err := root.Query().Proposal(ctx, &id, nil)
+		p := pn.(*protoTypes.GovernanceData)
 
 		t.Run("Proposal should be for a new spot market", func(t *testing.T) {
 			terms, err = root.Proposal().Terms(ctx, p)
@@ -810,7 +810,6 @@ func TestNewResolverRoot_Proposals(t *testing.T) {
 		)
 
 		var (
-			p         *protoTypes.GovernanceData
 			terms     *protoTypes.ProposalTerms
 			newMarket *protoTypes.ProposalTerms_NewMarket
 			asset     *protoTypes.Asset
@@ -818,7 +817,8 @@ func TestNewResolverRoot_Proposals(t *testing.T) {
 			err       error
 		)
 
-		p, err = root.Query().Proposal(ctx, &id, nil)
+		pn, err := root.Query().Proposal(ctx, &id, nil)
+		p := pn.(*protoTypes.GovernanceData)
 
 		t.Run("Proposal terms should be for a new market", func(t *testing.T) {
 			terms, err = root.Proposal().Terms(ctx, p)
@@ -870,14 +870,14 @@ func TestNewResolverRoot_Proposals(t *testing.T) {
 		)
 
 		var (
-			p         *protoTypes.GovernanceData
 			terms     *protoTypes.ProposalTerms
 			newMarket *protoTypes.ProposalTerms_UpdateMarket
 			product   *protoTypes.UpdateInstrumentConfiguration_Future
 			err       error
 		)
 
-		p, err = root.Query().Proposal(ctx, &id, nil)
+		pn, err := root.Query().Proposal(ctx, &id, nil)
+		p := pn.(*protoTypes.GovernanceData)
 
 		t.Run("Proposal terms should be to update market", func(t *testing.T) {
 			terms, err = root.Proposal().Terms(ctx, p)
@@ -954,13 +954,13 @@ func TestNewResolverRoot_Proposals(t *testing.T) {
 		)
 
 		var (
-			p         *protoTypes.GovernanceData
 			terms     *protoTypes.ProposalTerms
 			newMarket *protoTypes.ProposalTerms_UpdateSpotMarket
 			err       error
 		)
 
-		p, err = root.Query().Proposal(ctx, &id, nil)
+		pn, err := root.Query().Proposal(ctx, &id, nil)
+		p := pn.(*protoTypes.GovernanceData)
 
 		t.Run("Proposal should be to update a spot market", func(t *testing.T) {
 			terms, err = root.Proposal().Terms(ctx, p)
@@ -1019,14 +1019,14 @@ func TestNewResolverRoot_Proposals(t *testing.T) {
 		)
 
 		var (
-			p         *protoTypes.GovernanceData
 			terms     *protoTypes.ProposalTerms
 			newMarket *protoTypes.ProposalTerms_UpdateMarket
 			product   *protoTypes.UpdateInstrumentConfiguration_Perpetual
 			err       error
 		)
 
-		p, err = root.Query().Proposal(ctx, &id, nil)
+		pn, err := root.Query().Proposal(ctx, &id, nil)
+		p := pn.(*protoTypes.GovernanceData)
 
 		t.Run("Proposal terms should be to update market", func(t *testing.T) {
 			// Test the proposal resolver to make sure the terms and underlying changes are correct

--- a/datanode/gateway/graphql/schema.graphql
+++ b/datanode/gateway/graphql/schema.graphql
@@ -127,9 +127,9 @@ type Subscription {
     "ID of the party from which you want trades updates"
     partyId: ID
   ): [TradeUpdate!]
-  @deprecated(
-    reason: "Use tradesStream instead as it allows for filtering multiple markets and/or parties at once"
-  )
+    @deprecated(
+      reason: "Use tradesStream instead as it allows for filtering multiple markets and/or parties at once"
+    )
 
   "Subscribe to the trades updates"
   tradesStream(
@@ -632,7 +632,7 @@ type Query {
     "Type of the order"
     type: OrderType!
   ): OrderEstimate!
-  @deprecated(reason: "Use estimateFees and estimatePosition instead")
+    @deprecated(reason: "Use estimateFees and estimatePosition instead")
 
   "Return an estimation of the potential cost for a new order"
   estimateFees(
@@ -922,7 +922,6 @@ type Query {
     pagination: Pagination
   ): OrderConnection
 
-
   "List statistics about paid liquidity fees"
   paidLiquidityFees(
     "Optional market ID to filter for."
@@ -955,7 +954,7 @@ type Query {
     id: ID
     "Optionally, locate proposal by its reference. If ID is set, this parameter is ignored."
     reference: String
-  ): Proposal
+  ): ProposalNode
 
   "All governance proposals in the Vega network"
   proposalsConnection(
@@ -1248,7 +1247,7 @@ type Transfer {
 }
 
 union TransferKind =
-  OneOffTransfer
+    OneOffTransfer
   | RecurringTransfer
   | OneOffGovernanceTransfer
   | RecurringGovernanceTransfer
@@ -2448,9 +2447,9 @@ type Market {
     "Pagination information"
     pagination: Pagination
   ): TradeConnection
-  @deprecated(
-    reason: "Simplify and consolidate trades query and remove nesting. Use trades query instead"
-  )
+    @deprecated(
+      reason: "Simplify and consolidate trades query and remove nesting. Use trades query instead"
+    )
 
   "Current depth on the order book for this market"
   depth(
@@ -2492,7 +2491,7 @@ type Market {
     "Pagination information"
     pagination: Pagination
   ): LiquidityProvisionsConnection
-  @deprecated(reason: "Use liquidityProvisions instead")
+    @deprecated(reason: "Use liquidityProvisions instead")
 
   "Timestamps for state changes in the market"
   marketTimestamps: MarketTimestamps!
@@ -2505,7 +2504,7 @@ type Market {
 
   "Quadratic slippage factor is used to cap the slippage component of maintainence margin - it is applied to the square of the slippage volume"
   quadraticSlippageFactor: String!
-  @deprecated(reason: "This field will be removed in a future release")
+    @deprecated(reason: "This field will be removed in a future release")
 
   """
   Optional: Parent market ID. A market can be a successor to another market. If this market is a successor to a previous market,
@@ -2683,9 +2682,9 @@ type Party {
     "Pagination information"
     pagination: Pagination
   ): TradeConnection
-  @deprecated(
-    reason: "Simplify and consolidate trades query and remove nesting. Use trades query instead"
-  )
+    @deprecated(
+      reason: "Simplify and consolidate trades query and remove nesting. Use trades query instead"
+    )
 
   "Collateral accounts relating to a party"
   accountsConnection(
@@ -2701,7 +2700,7 @@ type Party {
 
   "Trading positions relating to a party"
   positionsConnection(market: ID, pagination: Pagination): PositionConnection
-  @deprecated(reason: "Use root positions query instead of sub-query")
+    @deprecated(reason: "Use root positions query instead of sub-query")
 
   "Margin levels for a market"
   marginsConnection(
@@ -2766,7 +2765,7 @@ type Party {
     "Optional Pagination information"
     pagination: Pagination
   ): LiquidityProvisionsConnection
-  @deprecated(reason: "Use liquidityProvisions instead")
+    @deprecated(reason: "Use liquidityProvisions instead")
 
   # All delegations for a party to a given node if node is specified, or all delegations if not
   delegationsConnection(
@@ -3105,9 +3104,9 @@ type Order {
     dateRange: DateRange
     pagination: Pagination
   ): TradeConnection
-  @deprecated(
-    reason: "Simplify and consolidate trades query and remove nesting. Use trades query instead"
-  )
+    @deprecated(
+      reason: "Simplify and consolidate trades query and remove nesting. Use trades query instead"
+    )
 
   "The order type"
   type: OrderType
@@ -4508,7 +4507,7 @@ type NewMarket {
   linearSlippageFactor: String!
   "Quadratic slippage factor is used to cap the slippage component of maintenance margin - it is applied to the square of the slippage volume"
   quadraticSlippageFactor: String!
-  @deprecated(reason: "This field will be removed in a future release")
+    @deprecated(reason: "This field will be removed in a future release")
   "Successor market configuration. If this proposed market is meant to succeed a given market, then this needs to be set."
   successorConfiguration: SuccessorConfiguration
   "Liquidity SLA Parameters"
@@ -4567,7 +4566,7 @@ type UpdateMarketConfiguration {
   linearSlippageFactor: String!
   "Quadratic slippage factor is used to cap the slippage component of maintenance margin - it is applied to the square of the slippage volume."
   quadraticSlippageFactor: String!
-  @deprecated(reason: "This field will be removed in a future release")
+    @deprecated(reason: "This field will be removed in a future release")
   "Liquidity SLA Parameters."
   liquiditySLAParameters: LiquiditySLAParameters
   "Specifies how the liquidity fee for the market will be calculated"
@@ -4619,7 +4618,7 @@ type UpdatePerpetualProduct {
 }
 
 union UpdateMarketRiskParameters =
-  UpdateMarketSimpleRiskModel
+    UpdateMarketSimpleRiskModel
   | UpdateMarketLogNormalRiskModel
 
 type UpdateMarketSimpleRiskModel {
@@ -4700,7 +4699,7 @@ type NewTransfer {
 }
 
 union GovernanceTransferKind =
-  OneOffGovernanceTransfer
+    OneOffGovernanceTransfer
   | RecurringGovernanceTransfer
 
 "Allows for cancellation of an existing governance transfer"
@@ -4861,7 +4860,7 @@ type NetworkParameter {
 }
 
 union ProposalChange =
-  NewMarket
+    NewMarket
   | UpdateMarket
   | UpdateNetworkParameter
   | NewAsset
@@ -6305,7 +6304,7 @@ type DataSourceSpecConfigurationTimeTrigger {
 }
 
 union InternalDataSourceKind =
-  DataSourceSpecConfigurationTime
+    DataSourceSpecConfigurationTime
   | DataSourceSpecConfigurationTimeTrigger
 
 """
@@ -6327,7 +6326,7 @@ type DataSourceDefinitionExternal {
 }
 
 union DataSourceKind =
-  DataSourceDefinitionInternal
+    DataSourceDefinitionInternal
   | DataSourceDefinitionExternal
 
 """
@@ -7029,7 +7028,7 @@ type GamesConnection {
 }
 
 "Connection type for retrieving cursor-based paginated party margin modes information"
-type  PartyMarginModesConnection {
+type PartyMarginModesConnection {
   "The party margin modes"
   edges: [PartyMarginModeEdge]
   "The pagination information"
@@ -7043,7 +7042,7 @@ type PartyMarginModeEdge {
 }
 
 "Margin mode selected for the given party and market."
-type PartyMarginMode{
+type PartyMarginMode {
   "Unique ID of the market."
   marketId: ID!
   "Unique ID of the party."

--- a/datanode/gateway/graphql/schema.graphql
+++ b/datanode/gateway/graphql/schema.graphql
@@ -6029,7 +6029,9 @@ union ProposalNode = Proposal | BatchProposal
 "Edge type containing the proposals and cursor information returned by a ProposalsConnection"
 type ProposalEdge {
   "The proposal data"
-  node: ProposalNode!
+  node: Proposal! @deprecated(reason: "Use proposalNode")
+  "The data of either single or batch proposal"
+  proposalNode: ProposalNode
   "Cursor identifying the proposal"
   cursor: String!
 }

--- a/datanode/gateway/graphql/schema.graphql
+++ b/datanode/gateway/graphql/schema.graphql
@@ -2423,7 +2423,7 @@ type Market {
   state: MarketState!
 
   "The proposal that initiated this market"
-  proposal: Proposal
+  proposal: ProposalNode
 
   "Orders on a market"
   ordersConnection(

--- a/datanode/gateway/graphql/schema.graphql
+++ b/datanode/gateway/graphql/schema.graphql
@@ -4913,6 +4913,31 @@ type ProposalTerms {
   change: ProposalChange!
 }
 
+"Terms for a batch governance proposal on Vega"
+type BatchProposalTermsChange {
+  """
+  RFC3339Nano time and date when this proposal is executed (if passed). Note that it has to be after closing date time.
+  Constrained by "minEnactInSeconds" and "maxEnactInSeconds" network parameters.
+  Note: Optional as free form proposals do not require it.
+  """
+  enactmentDatetime: Timestamp
+
+  "Actual change being introduced by the proposal - action the proposal triggers if passed and enacted."
+  change: ProposalChange!
+}
+
+"The rationale behind the proposal"
+type BatchProposalTerms {
+  """
+  RFC3339Nano time and date when voting closes for this proposal.
+  Constrained by "minClose" and "maxClose" network parameters.
+  """
+  closingDatetime: Timestamp!
+
+  "Actual changes being introduced by the proposal - actions the proposal triggers if passed and enacted."
+  changes: [BatchProposalTermsChange]!
+}
+
 """
 Various proposal types that are supported by Vega
 """
@@ -5000,9 +5025,42 @@ type Proposal {
   requiredLpParticipation: String
 }
 
+type BatchProposal {
+  "Proposal ID that is filled by Vega once proposal reaches the network"
+  id: ID
+  "A UUID reference to aid tracking proposals on Vega"
+  reference: String!
+  "Party that prepared the proposal"
+  party: Party!
+  "State of the proposal"
+  state: ProposalState!
+  "RFC3339Nano time and date when the proposal reached Vega network"
+  datetime: Timestamp!
+  "Rationale behind the proposal"
+  rationale: ProposalRationale!
+  "Votes cast for this proposal"
+  votes: ProposalVotes!
+  "Why the proposal was rejected by the core"
+  rejectionReason: ProposalRejectionReason
+  "Error details of the rejectionReason"
+  errorDetails: String
+  "Required majority for this proposal to succeed"
+  requiredMajority: String!
+  "Required participation for this proposal to succeed"
+  requiredParticipation: String!
+  "Equity-like share required for a market amendment proposal to be enacted (if not met, the proposal will not be enacted), represented as a fraction that can be converted to a percentage"
+  requiredLpMajority: String
+  "The market share of LPs' equity-like share that must take part in the market amendment vote for the proposal to pass. This means the votes of LPs that have submitted more liquidity to that market, or have been LPs from the start carry more weight. If it requires 50% of a market's equity-like share for a majority, and the proposal receives only YES votes but only LPs with 49% of the equity-like share voted, the proposal will not pass"
+  requiredLpParticipation: String
+  "Child proposals that are part of the batch"
+  subProposals: [ProposalDetail]
+}
+
 type ProposalDetail {
   "Proposal ID that is provided by Vega once proposal reaches the network"
   id: ID
+  "Parent proposal ID that is provided by Vega once proposal reaches the network"
+  batchId: ID
   "A UUID reference to aid tracking proposals on Vega"
   reference: String!
   "Party that prepared the proposal"
@@ -5011,8 +5069,10 @@ type ProposalDetail {
   state: ProposalState!
   "RFC3339Nano time and date when the proposal reached the Vega network"
   datetime: Timestamp!
-  "Terms of the proposal"
-  terms: ProposalTerms!
+  "Terms of the proposal - in case of single proposal"
+  terms: ProposalTerms
+  "Terms of the proposal - in case of batch proposal"
+  batchTerms: BatchProposalTerms
   "Rationale behind the proposal"
   rationale: ProposalRationale!
   "Why the proposal was rejected by the core"
@@ -5966,10 +6026,12 @@ enum TransferDirection {
   ToOrFrom
 }
 
+union ProposalNode = Proposal | BatchProposal
+
 "Edge type containing the proposals and cursor information returned by a ProposalsConnection"
 type ProposalEdge {
   "The proposal data"
-  node: Proposal!
+  node: ProposalNode!
   "Cursor identifying the proposal"
   cursor: String!
 }

--- a/datanode/gateway/graphql/schema.graphql
+++ b/datanode/gateway/graphql/schema.graphql
@@ -2423,7 +2423,10 @@ type Market {
   state: MarketState!
 
   "The proposal that initiated this market"
-  proposal: ProposalNode
+  proposal: Proposal @deprecated(reason: "Use marketProposal")
+
+  "The proposal that initiated this market"
+  marketProposal: ProposalNode
 
   "Orders on a market"
   ordersConnection(

--- a/datanode/gateway/graphql/schema.graphql
+++ b/datanode/gateway/graphql/schema.graphql
@@ -230,10 +230,10 @@ type PerpetualData {
   "Time at which the funding period started"
   startTime: Timestamp!
   "The index price used for external VWAP calculation"
-  indexPrice: String! 
-   "The methodology used to calculated index price for perps"
+  indexPrice: String!
+  "The methodology used to calculated index price for perps"
   indexPriceType: CompositePriceType!
-   "RFC3339Nano time indicating the next time index price will be calculated for perps where applicable"
+  "RFC3339Nano time indicating the next time index price will be calculated for perps where applicable"
   nextIndexPriceCalc: String!
 }
 
@@ -4912,12 +4912,11 @@ type ProposalTerms {
   change: ProposalChange!
 }
 
-"Terms for a batch governance proposal on Vega"
+"Terms for a batch governance proposal"
 type BatchProposalTermsChange {
   """
-  RFC3339Nano time and date when this proposal is executed (if passed). Note that it has to be after closing date time.
+  RFC3339Nano time and date when this proposal will be executed, if it passes.
   Constrained by "minEnactInSeconds" and "maxEnactInSeconds" network parameters.
-  Note: Optional as free form proposals do not require it.
   """
   enactmentDatetime: Timestamp
 
@@ -4925,7 +4924,7 @@ type BatchProposalTermsChange {
   change: ProposalChange!
 }
 
-"The rationale behind the proposal"
+"The rationale for the proposal"
 type BatchProposalTerms {
   """
   RFC3339Nano time and date when voting closes for this proposal.
@@ -5025,7 +5024,7 @@ type Proposal {
 }
 
 type BatchProposal {
-  "Proposal ID that is filled by Vega once proposal reaches the network"
+  "Proposal ID that is provided by Vega once proposal reaches the network"
   id: ID
   "A UUID reference to aid tracking proposals on Vega"
   reference: String!
@@ -5033,32 +5032,32 @@ type BatchProposal {
   party: Party!
   "State of the proposal"
   state: ProposalState!
-  "RFC3339Nano time and date when the proposal reached Vega network"
+  "RFC3339Nano time and date when the proposal reached the network"
   datetime: Timestamp!
   "Rationale behind the proposal"
   rationale: ProposalRationale!
   "Votes cast for this proposal"
   votes: ProposalVotes!
-  "Why the proposal was rejected by the core"
+  "Reason the proposal was rejected"
   rejectionReason: ProposalRejectionReason
-  "Error details of the rejectionReason"
+  "Details of the rejection reason"
   errorDetails: String
   "Required majority for this proposal to succeed"
   requiredMajority: String!
   "Required participation for this proposal to succeed"
   requiredParticipation: String!
-  "Equity-like share required for a market amendment proposal to be enacted (if not met, the proposal will not be enacted), represented as a fraction that can be converted to a percentage"
+  "Equity-like share required for a market amendment proposal to be enacted, represented as a fraction that can be converted to a percentage. If not met, the proposal will not be enacted"
   requiredLpMajority: String
-  "The market share of LPs' equity-like share that must take part in the market amendment vote for the proposal to pass. This means the votes of LPs that have submitted more liquidity to that market, or have been LPs from the start carry more weight. If it requires 50% of a market's equity-like share for a majority, and the proposal receives only YES votes but only LPs with 49% of the equity-like share voted, the proposal will not pass"
+  "The market share of LPs' equity-like share that must take part in a market amendment vote for the proposal to pass. This means the votes of LPs that have submitted more liquidity to that market, or have been LPs from the start carry more weight. If it requires 50% of a market's equity-like share for a majority, and the full batch of proposals receives all YES votes but only LPs with 49% of the equity-like share voted, the proposal will not pass"
   requiredLpParticipation: String
-  "Child proposals that are part of the batch"
+  "Proposals that are part of the batch"
   subProposals: [ProposalDetail]
 }
 
 type ProposalDetail {
   "Proposal ID that is provided by Vega once proposal reaches the network"
   id: ID
-  "Parent proposal ID that is provided by Vega once proposal reaches the network"
+  "Batch proposal ID that is provided by Vega once proposal reaches the network"
   batchId: ID
   "A UUID reference to aid tracking proposals on Vega"
   reference: String!
@@ -5068,9 +5067,9 @@ type ProposalDetail {
   state: ProposalState!
   "RFC3339Nano time and date when the proposal reached the Vega network"
   datetime: Timestamp!
-  "Terms of the proposal - in case of single proposal"
+  "Terms of the proposal for proposal"
   terms: ProposalTerms
-  "Terms of the proposal - in case of batch proposal"
+  "Terms of the proposal for a batch proposal"
   batchTerms: BatchProposalTerms
   "Rationale behind the proposal"
   rationale: ProposalRationale!

--- a/protos/vega/governance.go
+++ b/protos/vega/governance.go
@@ -4,3 +4,5 @@ type (
 	ProposalOneOffTermChangeType      = isProposalTerms_Change
 	ProposalOneOffTermBatchChangeType = isBatchProposalTermsChange_Change
 )
+
+func (gd *GovernanceData) IsProposalNode() {}

